File: extensions/select.py
# extensions/select.py
import debugutils
from state import save_selected
from wutildeps import windows
mark = debugutils.mark_time
log = debugutils.log
class Extension:
    def __init__(self):
        self.name = "select"
        self.desc = "Select a window by fuzzy title match."
        self.args = ["search_term"]
        self.short = "sel"
        self.requires_window = False
        self.deps = ["pywinctl"]
    def main(self, search_term):
        mark("select start", source="select")
        log(f"searching for window term={search_term}", source="select")
        all_windows = windows.all_windows()
        matches = windows.find_by_title(search_term)
        log(f"window candidates total={len(all_windows)} matches={len(matches)}", important=False, source="select")

        if not matches:
            print(f"No window found containing '{search_term}'.")
            log("select found no matches", source="select")
            return None

        if len(matches) > 1:
            print(f"Multiple matches for '{search_term}':")
            for i, w in enumerate(matches):
                print(f"[{i}] {w.title}")
            try:
                index = int(input("Select index: "))
                window = matches[index]
            except (ValueError, IndexError):
                print("Invalid choice.")
                log("select received invalid interactive choice", source="select")
                return None
        else:
            window = matches[0]

        print(f"Selected window: '{window.title}'")
        log(f"selected window title={window.title}", source="select")
        save_selected(window)
        mark("select complete", source="select")
        return window

File: extensions/select.go
package main

//wutil:name select
//wutil:short sel
//wutil:desc Select a window by fuzzy title match.
//wutil:args search_term
//wutil:requires_window false

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"

	"windowutil/pkg/wutilext"
	"windowutil/pkg/wutilwin"
)

func main() {
	search := ""
	if len(os.Args) > 1 {
		search = os.Args[1]
	}
	wutilext.Mark("select start", "select")
	wutilext.Log(fmt.Sprintf("searching for window term=%s", search), "select")

	all, _ := wutilwin.All()
	matches, err := wutilwin.FindByTitle(search)
	if err != nil {
		fmt.Printf("An error occurred: %s\n", err)
		wutilext.Log(fmt.Sprintf("select failed: %s", err), "select")
		return
	}
	wutilext.LogDetail(fmt.Sprintf("window candidates total=%d matches=%d", len(all), len(matches)), "select")
	if len(matches) == 0 {
		fmt.Printf("No window found containing '%s'.\n", search)
		wutilext.Log("select found no matches", "select")
		return
	}

	win := matches[0]
	if len(matches) > 1 {
		fmt.Printf("Multiple matches for '%s':\n", search)
		for i, match := range matches {
			fmt.Printf("[%d] %s\n", i, match.Title())
		}
		fmt.Print("Select index: ")
		line, _ := bufio.NewReader(os.Stdin).ReadString('\n')
		index, err := strconv.Atoi(strings.TrimSpace(line))
		if err != nil || index < 0 || index >= len(matches) {
			fmt.Println("Invalid choice.")
			wutilext.Log("select received invalid interactive choice", "select")
			return
		}
		win = matches[index]
	}

	if err := wutilwin.SaveSelected(win); err != nil {
		fmt.Printf("Failed to save selected window: %s\n", err)
		return
	}
	fmt.Printf("Selected window: '%s'\n", win.Title())
	wutilext.Log(fmt.Sprintf("selected window title=%s", win.Title()), "select")
	wutilext.Mark("select complete", "select")
}
